MAINT/ENH: Support for weights and range when estimating optimal number of bins

nayyarv opened the pull request:
Following on from PR #6029, the methods provided ignore range or weights so this PR aims to address this. 

The estimators are still defined as separate functions if there is ever a future need to expose these estimators to users. Each estimator now takes x and weights as arguments. These subfunctions expectation of `weights` is that it is count-like. If it's a probability, the subfunctions won't choose the datasize properly.

`_hist_optim_numbins_estimator` handles the range keyword by masking the appropriate parts of `a` and `weights`. It also checks whether weights is of an appropriate dtype (int or float). If weights is probability-like (i.e. `weights.sum()<a.size`), I scale weights such that `weights.sum()=a.size`. This means that I effectively choose  `n=max(a.size, weights.sum())` for my datasize, allowing for decent behaviour whether weights is count-like or probability like, while keeping things simple.

Freedman Diaconis requires a weighted IQR to be calculated, and since `np.percentile` or `np.partition` does not provide weighted options, this method is only used for unweighted data. Trying to use the 'FD' estimator with weights results in a `TypeError`. 'auto' now chooses between 'sturges' and 'scott' for weighted data, while retaining original behaviour for unweighted data. 

I've put some basic tests together and minor changes to the documentation and would love to hear others thoughts and feedback on the design choices/implementation/style etc.

tacaswell committed:
BUG: many functions silently drop `keepdims` kwarg

change test from `type(a) is not mu.ndarray` to
`not isinstance(a, mu.ndarray)`

Because every sub-class of ndarray is not guaranteed to implement
`keepdims` as a kwarg, when wrapping these methods care must be taken.

The previous behavior was to silently eat the kwarg when dealing with
a sub-class of ndarray.

Now, if `keepdims=np._NoValue` (the new default) it is not passed
through to the underlying function call (so the default value of
`keepdims` is now controlled by the sub-class).  If `keepdims` is not
`np._NoValue` then it is passed through and will raise an exception if
the sub-class does not support the kwarg.

A special case in nanvar was required to deal with `matrix` that previously
relied on `fromnumeric` silently dropping `keepdims`.

Commit:
BUG: Fix nanpercentile crash on all-nan slices

Fix bug where nanpercentile would crash with an all-nan slices when
given multiple percentiles.  Also corrects behavior where array sizes
different from numpy.percentile would be returned with keepdims enabled.

Fix #5760

shoyer commented on numpy/lib/function_base.py:
```diff
@@ -91,16 +102,37 @@ def _hist_optim_numbins_estimator(a, estimator):
+        mn, mx = data_range
+        keep = (a >= mn)
+        keep &= (a <= mx)
+        if not np.logical_and.reduce(keep):
```
use `np.all` instead

nayyarv replied:
I coped the filter step from https://github.com/numpy/numpy/blob/master/numpy/lib/function_base.py#L400 which came up in #6100. 

It's actually faster than `np.all`

shoyer commented on numpy/lib/function_base.py:
```diff
@@ -91,16 +102,37 @@ def _hist_optim_numbins_estimator(a, estimator):
+                data_weights = data_weights[keep]
+
+    if data_weights is not None:
+        if data_weights.dtype != int and data_weights.dtype != float:
```
probably better to use `np.issubdtype` instead to allow for, e.g., float32

shoyer commented on numpy/lib/function_base.py:
```diff
@@ -91,16 +102,37 @@ def _hist_optim_numbins_estimator(a, estimator):
+            # we assume that weights represents counts. In the case that weights
+            # represents a probability (i.e. weights.sum()<=1),
+            # we scale the weights to represent a count where weights.sum = a.size
+            data_weights = data_weights * (a.size/data_weights.sum())
```
Why not simply set `data_size = np.maximum(a.size, weights.sum())` directly? I don't think there is anything else which needs a rescaled `weights`

shoyer commented on numpy/lib/function_base.py:
```diff
@@ -141,14 +186,14 @@ def fd(x):
         This method gives good off the shelf behaviour.
         """
-        return max(fd(x), sturges(x))
+        return max(scott(x, weights), sturges(x, weights)) if weights is not None else max(fd(x), sturges(x))
```
this is too much logic for one line

shoyer replied:
also, it's worth writing a comment with the reasoning for the switch

shoyer commented:
The design choices here look good to me.

seberg commented:
Frankly, I am a bit sceptical about automatically guessing the weight type. Maybe it is clear enough, just wanted to say that it has a bit of a bad taste to me for starters.

That said, I think the functionality is not bad at all.

nayyarv commented:
I agree that choosing data_size from weights is far from perfect, but without further input from the user (`**kwargs` or another argument in `histogram`)/exposing the estimators, it's the best I can do.
After our discussion at the end of the last PR, we decided the two main use cases for weights + automatic bin estimation were either count-like or probability-like weights (matched up to friends/colleagues thoughts)
If the weights were count-like, it's sum would be greater than x.size and if it was probability-like, it's sum would be less, which is why I use the `max(a.size, weights.sum())` logic. 
If the user was using the weights option as a bin summation method, it's unlikely they would use automatic estimators, and rather fixed bin edges - so situations in which the weights are not count/probability like can be ignored. 

Any better solutions, I'm all ears. If nothing else, we could simply patch in the `range` support and deal with weights at a later date.

seberg commented:
I am not currently convinced that you can always correctly infer the type. So I am still wondering if we cannot find a way to force the user to make a conscies choice.

nayyarv commented:
Fair enough - the way I see it, we need to know 2 things
1. Whether or not to consider the weights when estimating the number of bins
2. If we do consider the weights, what should the data size be - `weights.sum()` or `a.size`. 

Basically, we need to know how to deal with datasize - if it's None, we can ignore the weights when estimating, if it's 'sum' or 'size', we know that we have to deal with weights. I.e. 1 argument.

Possible solutions,
1. Expose the estimators for manual control and let `np.histogram` provide simple default behaviour where it tries to guess (Maybe move the estimators to `scipy.stats`?)
2. Add another argument to `np.histogram` for instructions on how to deal with weights, `weights_sum=True` or something.
3. Add options to the `bins` string, like `'auto sum'` or `'scott size'`, which can then be `split` to retrieve the estimator and how to deal with weights. If the second term isn't included, ignore weights in the estimation. This also means the `bins` checking logic remains simple, as opposed to making it a tuple or list which is also iterable.
4. Something similar to 3, except maybe combined into the `weights` argument as a tuple or iterable?

I'm in favour of 3 (or maybe 1), as it requires the least amount of API changes, and it's an incremental change.

njsmith commented:
Would it make sense to add a new kwarg data_size whose value defaults to
weights.sum()?
On Sep 9, 2015 1:01 AM, "Varun Nayyar" <notifications@github.com> wrote:

> Fair enough - the way I see it, we need to know 2 things
> 1. Whether or not to consider the weights when estimating the number of
> bins
> 2. If we do consider the weights, what should the data size be -
> weights.sum() or a.size.
>
> Basically, we need to know how to deal with datasize - if it's None, we
> can ignore the weights when estimating, if it's 'sum' or 'size', we know
> that we have to deal with weights. I.e. 1 argument.
>
> Possible solutions,
> 1. Expose the estimators for manual control and let np.histogram provide
> simple default behaviour where it tries to guess (Maybe move the estimators
> to scipy.stats?)
> 2. Add another argument to np.histogram for instructions on how to deal
> with weights, weights_sum=True or something.
> 3. Add options to the bins string, like 'auto sum' or 'scott size', which
> can then be split to retrieve the estimator and how to deal with weights.
> If the second term isn't included, ignore weights in the estimation. This
> also means the bins checking logic remains simple, as opposed to making
> it a tuple or list which is also iterable.
> 4. Something similar to 3, except maybe combined into the weights
> argument as a tuple or iterable?
>
> I'm in favour of 3 (or maybe 1), as it requires the least amount of API
> changes, and it's an incremental change.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numpy/numpy/pull/6288#issuecomment-138817995>.
>

nayyarv commented:
I'm not overly keen on adding a new kwarg that only applies when a bins estimator is being used and `weights` is not None. Seems like overkill, and I would like to avoid R style functions with their mostly unused arguments.

njsmith commented:
I'm not keen on having lots of non-orthogonal kwargs either, but I like it
a lot better than the idea of defining a little ad hoc string language
inside a single string kwarg. The other general option that comes to mind
would be using something that has more structure than a str, like a dict or
the object, but I'm not seeing a terribly natural way to apply that trick
here myself...
On Sep 9, 2015 6:31 AM, "Varun Nayyar" <notifications@github.com> wrote:

> I'm not overly keen on adding a new kwarg that only applies when a bins
> estimator is being used and weights is not None. Seems like overkill, and
> I would like to avoid R style functions with their mostly unused arguments.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numpy/numpy/pull/6288#issuecomment-138909313>.
>

tacaswell commented:
If there are only two options for how to deal with the weights, maybe just do a look up on the whole string.  You effectively double/triple  the size of the number of estimators, but that does not seem so bad.  

From the mpl side, having `histogram` gain any kwargs would greatly complicate the argument handling in `hist`.

shoyer commented:
I wonder if it would make sense to pick `data_size` based on the dtype of the array:

- int: use `data.sum()`
- float: use `data.size`

I guess this would interact poorly with downstream libraries like pandas that use floats to represent nullable-int.

njsmith commented:
Is using data.size even theoretically justified? I can't quite think of any
reason why it's a more natural choice than any other arbitrary number
(though I admit my intuition about these algorithms is very limited).

On Wed, Sep 9, 2015 at 11:41 PM, Stephan Hoyer <notifications@github.com>
wrote:

> I wonder if it would make sense to pick data_size based on the dtype of
> the array:
>
>    - int: use data.sum()
>    - float: use data.size
>
> I guess this would interact poorly with downstream libraries like pandas
> that use floats to represent nullable-int.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numpy/numpy/pull/6288#issuecomment-139131087>.
>



-- 
Nathaniel J. Smith -- http://vorpus.org

nayyarv commented:
Haha, inferring `data_size` from the dtype is what we're unhappy with, so we're looking at a way to get the user to give us input without changing function signatures too much.

The estimator is primarily concerned with the size of the data to minimise Mean Integrated Squared Error. In general optimal binwidth ~ n^(-1/3) where n is the datasize.

Possible situations
1. Now if weights represents counts, this means that `np.repeat(a, weights)` would give us the full data set and hence n should be `weights.sum()`. This makes sense for integer weights and some float weights (if there's reasoning behind it).
2.In a situation where the weights represents a probability/proportion, the calculation of standard deviation or the IQR for scott's and FD's rule respectively are now altered, while the datasize should be the number of points, not `weight.sum()` which would evaluate to 1.
3. Finally, if there is no meaning to the weights, it's simply for the summation, the weights should be ignored for the calculation of the standard deviation and the datasize. This is for something exotic like complex weights or you have matched pairs eg. (Age, Wealth)  that you want to sum up per group.

We could potentially turn `bins` into a dict (or a list or tuple) for automatic methods, i.e,
`{method='auto', weights=['sum', 'size', 'ignore']}` for example, and for `bins='auto'` we either ignore weights by default or try and guess (`max(weights.sum(), a.size)` for data_size) for best average case behaviour.

nbeaver committed:
BUG: check lower limit of base in base_repr.

seberg commented:
Yeah, the whole discussion is because I am unhappy with guessing what to
do with weights based on the type. I have the feeling people are not
necessary enough aware of their type. And accuracy weights might add up
to more then the sample size in some cases as well....

seberg commented:
I would like to throw in one further option, which may or may not be temporary:

Ignore the problem ;), maybe by exposing the bin number functions. Then just throw an error and tell user to use those functions manually. If weights are not very common (I have no clue) for plotting, we could do that reasonably.

tacaswell commented:
I like @seberg's last suggestion (ignore/raise) the best.

On Thu, Sep 10, 2015 at 7:08 AM seberg <notifications@github.com> wrote:

> I would like to throw in one further option, which may or may not be
> temporary:
>
> Ignore the problem ;), maybe by exposing the bin number functions. Then
> just throw an error and tell user to use those functions manually. If
> weights are not very common (I have no clue) for plotting, we could do that
> reasonably.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numpy/numpy/pull/6288#issuecomment-139204884>.
>

shoyer commented:
So just raising an error is weights are set? That seems reasonable to me, better than designing an awkward API for niche uses.

On Thu, Sep 10, 2015 at 10:18 AM, Thomas A Caswell
<notifications@github.com> wrote:

> I like @seberg's last suggestion (ignore/raise) the best.
> On Thu, Sep 10, 2015 at 7:08 AM seberg <notifications@github.com> wrote:
>> I would like to throw in one further option, which may or may not be
>> temporary:
>>
>> Ignore the problem ;), maybe by exposing the bin number functions. Then
>> just throw an error and tell user to use those functions manually. If
>> weights are not very common (I have no clue) for plotting, we could do that
>> reasonably.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/numpy/numpy/pull/6288#issuecomment-139204884>.
>>
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/numpy/numpy/pull/6288#issuecomment-139315898

njsmith commented:
It would be nice to support count weights + auto binning at some point, but
we can always do that as an enhancement later once we've figured out how to
actually make it work right :-). And until then, raising an error is safe
way to keep our options open, so +1

On Thu, Sep 10, 2015 at 11:08 AM, Stephan Hoyer <notifications@github.com>
wrote:

> So just raising an error is weights are set? That seems reasonable to me,
> better than designing an awkward API for niche uses.
>
> On Thu, Sep 10, 2015 at 10:18 AM, Thomas A Caswell
> <notifications@github.com> wrote:
>
> > I like @seberg's last suggestion (ignore/raise) the best.
> > On Thu, Sep 10, 2015 at 7:08 AM seberg <notifications@github.com> wrote:
> >> I would like to throw in one further option, which may or may not be
> >> temporary:
> >>
> >> Ignore the problem ;), maybe by exposing the bin number functions. Then
> >> just throw an error and tell user to use those functions manually. If
> >> weights are not very common (I have no clue) for plotting, we could do
> that
> >> reasonably.
> >>
> >> —
> >> Reply to this email directly or view it on GitHub
> >> <https://github.com/numpy/numpy/pull/6288#issuecomment-139204884>.
> >>
> > ---
> > Reply to this email directly or view it on GitHub:
> > https://github.com/numpy/numpy/pull/6288#issuecomment-139315898
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numpy/numpy/pull/6288#issuecomment-139330477>.
>



-- 
Nathaniel J. Smith -- http://vorpus.org

nayyarv commented:
Ok, in that case, I'll edit the PR to support the `range` keyword only and throw an error for `weights not None`. R doesn't provide an option for `weights` with it's `hist` method (which is primarily visual), and I'm not sure how many situations with `'auto'` bin estimation will have weights too. I suppose we'll find out after 1.11 is released.

With regards to exposing the estimators, would it find a better home in `scipy.stats` instead?

seberg commented:
It is true that exposing would fit better into scipy. And I am not a fan of exposing them in top level numpy. But scipy is not a dependency so that would be code duplication.
So, I know I suggested that we may expose it, but I frankly have no clue how it would look like :(.

nayyarv committed:
ENH: Adding support to the range keyword for estimation of the optimal number of bins and associated tests

nayyarv commented:
Removed support for weights, and instead raise a TypeError when weights!=None is passed into the function. Support for range still remains, and the test is included too.

nayyarv commented on numpy/lib/function_base.py:
```diff
@@ -91,6 +101,17 @@ def _hist_optim_numbins_estimator(a, estimator):
         return 1
 
+    if data_weights is not None:
+        raise TypeError("Automated estimation of the number of "
```
Maybe a ValueError? What's the best error to raise here?

madphysicist replied:
How about a `RuntimeWarning`? Something along the lines of `FD estimator currently ignores weighted data`. Might be better than crashing.

jonathanunderwood committed:
ENH: Allow specification of terms to fit in legfit

The argument `deg` is enhanced to allow an
array_like argument to past which specifies
which terms to include in the fit.

The returned coef array is exapnded to have
entries of 0 for all coefficients which were
not included in the fit.

jonathanunderwood committed:
TST: Add tests for legfit with deg specified as list

jonathanunderwood committed:
ENH: Allow specification of terms to fit in chebfit

The argument `deg` is enhanced to allow an
array_like argument to past which specifies
which terms to include in the fit.

The returned coef array is exapnded to have
entries of 0 for all coefficients which were
not included in the fit.

jonathanunderwood committed:
TST: Add tests for chebfit with deg specified as list

jonathanunderwood committed:
ENH: Allow specification of terms to fit in hermfit

The argument `deg` is enhanced to allow an
array_like argument to past which specifies
which terms to include in the fit.

The returned coef array is exapnded to have
entries of 0 for all coefficients which were
not included in the fit.

jonathanunderwood committed:
TST: Add tests for hermfit with deg specified as list

stuarteberg committed:
MAINT: Add '0x' to API version error for clarity.

jonathanunderwood committed:
ENH: Allow specification of terms to fit in lagfit

The argument `deg` is enhanced to allow an
array_like argument to past which specifies
which terms to include in the fit.

The returned coef array is exapnded to have
entries of 0 for all coefficients which were
not included in the fit.

jonathanunderwood committed:
TST: Add tests for lagfit with deg specified as list

I--P committed:
ENH: usecols now accepts an int when only one column has to be read

Commit:
BUG: fix issues with signed zeros in scalar math complex division.

The current algorithm used in scalar math complex division appears to
incorrectly handle signed zeros. This patch duplicates the algorithm
used for complex division in the loops.c.src file into the
scalarmath.c.src file so the algorithms are consistent regardless of
context. Unit tests are added in the scalar context for testing the
correctness of sign when zeros are encountered and also to trip the new
branches in the now consistent algorithm.

jonathanunderwood committed:
ENH: Allow specification of terms to fit in polyfit

The argument `deg` is enhanced to allow an
array_like argument to past which specifies
which terms to include in the fit.

The returned coef array is exapnded to have
entries of 0 for all coefficients which were
not included in the fit.

jonathanunderwood committed:
TST: Add tests for polyfit with deg specified as list

jonathanunderwood committed:
ENH: Allow specification of terms to fit in fit method

jonathanunderwood committed:
TST: Add tests for check_fit with deg specified as list

jonathanunderwood committed:
ENH: Allow specification of terms to fit in hermefit

The argument `deg` is enhanced to allow an
array_like argument to past which specifies
which terms to include in the fit.

The returned coef array is exapnded to have
entries of 0 for all coefficients which were
not included in the fit.

jonathanunderwood committed:
TST: Add tests for hermefit with deg specified as list

anntzer committed:
MAINT: memcpy-based ~4x faster, typed shuffle.

Only for 1d-ndarrays exactly, as subtypes (e.g. masked arrays) may not
allow direct shuffle of the underlying buffer (in fact, the old
implementation destroyed the underlying values of masked arrays while
shuffling).

Also handles struct-containing-object 1d ndarrays properly.

See #6776 for an earlier, less general (but even faster: ~6x)
improvement attempt, #5514 for the original issue.

rgommers committed:
BLD: require setuptools for Numpy builds.  Closes gh-6551.

rgommers committed:
MAINT: remove unnecessary setupegg.py file, now that we depend on setuptools.

rgommers committed:
BLD: disable (half-)broken setuptools commands and give clear error messages.

rgommers committed:
BUG: make distutils.misc_util.is_string recognize unicode.

This was triggered by the numpy version string (which is unicde)
being included in metadata.  This could also solve other problems
for external users of numpy.distutils.

rgommers committed:
BLD: finish handling of setuptools commands.

Also ignore setup.cfg: this file is created/modified by the
alias/setopt/saveopts commands, and therefore needs to be in .gitignore.

rgommers committed:
BUG: fix TravisCI test issues when using setuptools unconditionally.

Also remove all mentions of setupegg.py from the documentation.

rgommers committed:
DOC: add Cython to list of dependencies in INSTALL.txt.  Closes gh-6682.

rgommers committed:
TST: add test to check for correct version string format.

Implements idea suggested in gh-6431.

rgommers committed:
DOC: add NumpyVersion to the docs, and mention it in version.py

Commit:
BUG: Enforce order param for MaskedArray construction

Adds the 'order' parameter to the __new__ override
in MaskedArray construction, enabling it to be enforced
in methods like np.ma.core.array and np.ma.core.asarray.

Closes gh-6646.

anntzer committed:
Top shuffle speed for machine-sized ints/floats.

Apparently gcc only specializes one branch (the last one) so I went for
another 33% performance increase (matching #6776) in what's likely the
most common use case.

jakirkham committed:
MAINT: Refactor `cblas_innerproduct` to use `cblas_matrixproduct`.

jakirkham committed:
MAINT: Refactor `PyArray_InnerProduct` so that it just performs a transpose and calls `PyArray_MatrixProduct2`.

jakirkham committed:
TST: Ensure `dot` fails correctly if array types cannot be coerced into a common type.

jakirkham committed:
MAINT: Ensure `inner` is raising a ValueError just as `dot` does in the same case.

jakirkham committed:
TST: Ensure `inner` fails correctly if array types cannot be coerced into a common type.

jakirkham committed:
BUG: Clear error before constructing error message using calls to `PyObject_Repr`. Also, do a better job of handling any errors raised while constructing the error message.

charris committed:
Merge pull request #6994 from juliantaylor/signbit

BUG: make result of isfinite/isinf/signbit a boolean

Commit:
MAINT: Simplified mtrand.pyx helpers

Refactored methods that broadcast arguments
together by finding additional common ground
between code in the if...else branches that
involved a size parameter being passed in.

charris committed:
Merge pull request #6968 from jakirkham/optimize_innerproduct

ENH, MAINT: Refactor `PyArray_InnerProduct` to use `PyArray_MatrixProduct2`

charris committed:
Merge pull request #6969 from charris/bench-randint

ENH: Add benchmark tests for numpy.random.randint.

charris committed:
Merge pull request #6975 from charris/add-cacosh-to-glibc-blacklist

BUG: Add more complex trig functions to glibc < 2.16 blacklist.

Commit:
DOC, MAINT: Enforce np.ndarray arg for np.put and np.place

np.put and np.place do something only when the first argument
is an instance of np.ndarray. These changes will cause a TypeError
to be thrown in either function should that requirement not be
satisfied.

ahaldane committed:
Merge pull request #6728 from gerritholl/structured_multidim_masked_array_fillvalue

BUG/TST: Fix for #6723 including test: force fill_value.ndim==0

anntzer committed:
Revert to non-generative test.

shoyer committed:
TST, ENH: make all comparisons with NaT false

Now, NaT compares like NaN:
- NaT != NaT -> True
- NaT == NaT (and all other comparisons) -> False

We discussed this on the mailing list back in October:
https://mail.scipy.org/pipermail/numpy-discussion/2015-October/073968.html

juliantaylor committed:
BUG: skip invalid path distutils warning for empty strings

empty strings are the default for the new rpath,
extra_compile_args and extra_link_args sections

rgommers committed:
Merge pull request #7005 from juliantaylor/get_path_warn

BUG: skip invalid path distutils warning for empty strings

TwistedHardware committed:
Fix number sequence

I have found that there are two missing numbers in a sequence in the documentation.
http://docs.scipy.org/doc/numpy/user/misc.html#interfacing-to-c

It goes 1,2,3,5,7,8 with missing 4 and 6.

jaimefrio committed:
Merge pull request #7008 from TwistedHardware/master

DOC: Fix a number sequence in docs

charris committed:
Merge pull request #6465 from shoyer/datetime64-NaT-casting

BUG: fix casting rules for generic datetime64/timedelta64 dtypes

charris committed:
Merge pull request #7001 from shoyer/NaT-comparison

API: make all comparisons with NaT false

njsmith committed:
DOC: Clean up/fix several references to the "future" 1.10 release

Fixes gh-7010

charris committed:
Merge pull request #7015 from njsmith/fix-1.10-messages

DOC: Clean up/fix several references to the "future" 1.10 release

charris committed:
Merge pull request #6912 from gfyoung/asarray_order

BUG: Enforce Order Param for MaskedArray Construction

charris committed:
TST: Add tests for divmod

charris committed:
BUG: Make divmod behave better under roundoff error.

This is apropos #6127. The fix is to make the functions floor_division
and remainder consistent, i.e.,

    b * floor_division(a, b) + remainder(a, b) == a

Previous to this fix remainder was computed a the C level using the '%'
operator, and the result was not always consistent with the floor
function. The current approach is to compute the remainder using

    b * (a/b - floor(a/b))

which is both consistent with the Python '%' operator and numerically
consistent with floor_division implemented using the floor function.

Closes #6127.

andyfaff committed:
MAINT: ensureisclose returns scalar when called with two scalars

shoyer committed:
TST: Make assert_warns an optional contextmanager

shoyer committed:
API: Make datetime64 timezone naive

Fixes GH3290

With apologies to mwiebe, this rips out most of the time zone parsing from
the datetime64 type.

I think we mostly sorted out the API design in discussions last year, but
I'll be posting this to the mailing list shortly to get feedback.

Old behavior:

	# string parsing and printing defaults to your local timezone :(
    >>> np.datetime64('2000-01-01T00')
    numpy.datetime64('2000-01-01T00:00-0800','h')

New behavior:

    # datetime64 is parsed and printed as timezone naive
    >>> np.datetime64('2000-01-01T00')
    numpy.datetime64('2000-01-01T00','h')

    # you can still supply a timezone, but you get a deprecation warning
    >>> np.datetime64('2000-01-01T00Z')
    DeprecationWarning: parsing timezone aware datetimes is deprecated; this
    will raise an error in the future
    numpy.datetime64('2000-01-01T00','h')

jakirkham committed:
DOC: Explain the new exception behavior of `np.dot` when its types cannot be cast to a common type.

charris committed:
Merge pull request #7016 from andyfaff/isclose

MAINT: ensure isclose returns scalar when called with two scalars closes #7014

jakirkham committed:
DOC: Fix markdown style inline code to restructured text style inline code.

jakirkham committed:
DEP: Add warnings to `__getitem__` and `__setitem__` to point out the behavior of `MaskedArray`'s masks is changing.

jakirkham committed:
TEST: Ignore `FutureWarning` if raised from running masked array operations.

jakirkham committed:
DOC: Explain that `MaskedArray`s will try to consistently return view of their masks when they are also returning views of their data.

shoyer committed:
Merge pull request #7000 from gfyoung/ndarray_arg_enforce

DOC, MAINT: Enforce np.ndarray arg for np.put and np.place

charris committed:
Merge pull request #6987 from jakirkham/test_dot_inner_type_failures

TST, MAINT: Make sure exceptions of `inner` and `dot` match for different cases

charris committed:
Merge pull request #7023 from jakirkham/fix_docs_release

DOC: Release 1.11.0 text `code` -> ``code``

Commit:
DOC: Clarify behavior in np.random.uniform

Although the arguments are specified as 'high'
and 'low', it is possible to pass in values for
'low' and 'high' where 'low' >= 'high' and still
obtain well-defined behavior. The documentation
has been expanded to reflect this fact, with a
note to discourage passing in arguments satisfying
'low' > 'high'.

charris committed:
Merge pull request #7020 from jakirkham/deprecated_masked_array_mask_copy

DEP: Warn MaskedArray will return views of mask when sliced

Commit:
TST: Added lots of new tests for fromnumeric.py

rgommers committed:
TST: fix issues with test for correctness of numpy version string.

Addresses comments of @pv on gh-6895.

Commit:
Fix issue 7021

rgommers committed:
DOC: update INSTALL.txt.

rgommers committed:
DOC: some tweaks to the install and build info in the user guide.

rgommers committed:
DOC: some more cleanup in INSTALL.txt, and rename to INSTALL.rst.txt

rgommers committed:
BLD: fix rebuilding after a failed build.  Closes gh-5467.

Also remove bench(), does't do anything here after the move to asv.

rgommers committed:
DOC/BUG: textual improvements in install docs, and fix a typo in setup.py

Address comments of @charris on gh-6895.

charris committed:
Merge pull request #6895 from rgommers/packaging

BLD: build improvements (enable setuptools)

charris committed:
Merge pull request #7029 from rgommers/fix-interrupted-rebuild

BLD: fix rebuilding after a failed build.  Closes gh-5467.

charris committed:
Merge pull request #7028 from sam09/fix-7021

Fix issue #7021

charris committed:
REL: Do not include merge commits in Changelog.

Cleans up the Changelog.

[ci skip]

ogrisel committed:
BLD: build travis dev wheels for py27

jakirkham committed:
BENCH: Reorganize existing benchmarks by the order they show up when run in the benchmarking suite.

jakirkham committed:
BENCH: Add some benchmarks for `matmul`.

jakirkham committed:
DOC: Update the release notes to state that the `A.T @ A` optimization has been extended to several NumPy operations.

rgommers committed:
Merge pull request #7032 from charris/no-merges-in-changelog

REL: Do not include merge commits in Changelog.

charris committed:
Merge pull request #7034 from jakirkham/bench_matmul

BENCH, DOC: Benchmark matmul and update documentation

jakirkham committed:
DOC: Fix typo.

charris committed:
Merge pull request #7022 from charris/fix-divmod

BUG: Fix divmod

charris committed:
Merge pull request #7031 from charris/cleanup-6699

Cleanup 6699, BUG: fix issues with signed zeros in scalar math complex division.

jakirkham committed:
STY: Wrap some long lines.

njsmith committed:
Merge pull request #7037 from jakirkham/1_10_4_rel_doc_fix

DOC: Fix typo

charris committed:
Merge pull request #7033 from ogrisel/py27-travis-wheels

[MRG] Enable py27 travis wheels upload for downstream projects

rgommers committed:
BLD: fix runtests.py, was broken by the move to setuptools.

Issue was introduced by gh-6895.

rgommers committed:
DOC: add build system changes to 1.11.0 release notes.

[ci skip]

shoyer committed:
DOC: document changes to assert_warns

charris committed:
Merge pull request #7040 from rgommers/fix-runtests-setuptools

BLD: fix runtests.py, was broken by the move to setuptools.

charris committed:
Revert "Merge pull request #7001 from shoyer/NaT-comparison"

This reverts commit 7141f40b58ed1e7071cde78ab7bc8ab37e9c5983,
reversing changes made to 8fa6e3bef26a6d4a2c92f2824129aa4409be2590.

The original broke some pandas tests. The current plan to get this
in is

* reversion
* issue FutureWarning in 1.11 and 1.12
* make the change in 1.13.

charris committed:
Merge pull request #7038 from jakirkham/cblasfuncs_sty_fixes

STY: Wrap some long lines in `cblasfuncs.c`

charris committed:
Merge pull request #6453 from shoyer/naive-datetime64

API: Make datetime64 timezone naive

charris committed:
Merge pull request #7042 from charris/revert-7001

Revert "Merge pull request #7001 from shoyer/NaT-comparison"

njsmith committed:
Merge pull request #6933 from anntzer/fastshuffle-memcpy

memcpy-based fast, typed shuffle.

juliantaylor committed:
Merge pull request #5981 from dfreese/fix/multi_percent_nanperc_bug

BUG: Handle multiple percentiles for all-nan slices in nanpercentile

rgommers committed:
TST: de-uglify the setuptools appeasement in runtests.py a bit.

This is a cleaner fix than the one committed in gh-7040.

charris committed:
Merge pull request #7045 from rgommers/runtests-no-eggs

TST: de-uglify the setuptools appeasement in runtests.py a bit.

zerothi committed:
ENH: Allow site.cfg information with libraries key

This PR fixes the case when users create a site.cfg to fix
library locations, but does not change the library names.

Now numpy.distutils correctly checks all options related to
libraries by defaulting to the library from the class via _lib_names

Signed-off-by: Nick Papior <nickpapior@gmail.com>

charris committed:
Merge pull request #6608 from stuarteberg/fix-api-version-warning-ints

MAINT: Add '0x' to API version error for clarity.

charris committed:
Merge pull request #7048 from zerothi/enh-fix-distutils.libs

ENH: Allow site.cfg information with libraries key

andyfaff committed:
TST: np.isclose, correct error msg on test

charris committed:
Merge pull request #7049 from andyfaff/isclosetst

TST: np.isclose, correct error msg on test

jonathanunderwood committed:
Add release note for polynomial fit deg changes

charris committed:
Merge pull request #6360 from jonathanunderwood/legfit_restrict_terms

ENH: Allow specification of terms to fit in polynomial fitting functions

charris committed:
MAINT: Refactor polynomial leastsquares fits.

* Rewrite the documentation for `deg`.
* Check that the passed degrees are int and raise if not.
* Do not accept `deg` arguments that do not convert as 0-D or 1-D.
* Sort passed degree terms for repeatability.

charris committed:
DOC: Update the 1.10.0 release notes.

Rewrite the documentation of added functionality of the deg parameter of
the polynomial fitting functions in the numpy.polynomial package and put
it in the 'New Features' section.

tushar-rishav committed:
BUG:Should fix astype cast bug in numpy.fft module
Closes #6693

Commit:
BUG: One element array inputs get one element arrays returned in np.random

Fixes bug in np.random methods that would return scalars
when passed one-element array inputs. This is because
one-element ndarrays can be cast to integers / floats, which
is what functions like PyFloat_AsDouble do before converting
to the intended data type.

This commit changes the check used to determine whether the
inputs are purely scalar by converting all inputs to arrays
and checking if the resulting shape is an empty tuple (scalar)
or not (array).

Closes gh-4263.

charris committed:
DEP: Emit FutureWarning for NAT comparisons.

In Numpy 1.13 the plan is for NAT comparisons to behave like NaN
comparisons, e.g., False except for 'NAT != NAT', which will be
True. See the discussion at gh-7019 for details.

charris committed:
TST: Add tests for NAT comparison FutureWarning.

The behavior of NAT comparisons will change in Numpy 1.13. Make sure
that a FutureWarning is emitted when the results will change.

charris committed:
Merge pull request #7057 from charris/nat-future-warning

DEP: Emit FutureWarning for NAT comparisons.

charris committed:
Merge pull request #7053 from tushar-rishav/ifft_RAM_requirements

BUG: Should fix astype cast bug in numpy.fft module

charris committed:
Merge pull request #7052 from charris/fixup-polynomials

Refactor functionality added to the numpy.polynomial package in #6360.

charris committed:
Merge pull request #7027 from gfyoung/test_fromnumeric_expand

TST, MAINT: Lots of new tests for fromnumeric.py

charris committed:
REL: Prepare for 1.10.x branch.

* Add comment to cversions.txt (no change)
* Add comment to setup_common.py (no change)
* Nothing done for numpy/core/include/numpy/numpyconfig.h
* Update log start to 1.10.0 in pavement.py.

charris committed:
Merge pull request #7062 from charris/pre-1.10.x-branch

REL: Prepare for 1.10.x branch.

charris committed:
REL: Update master branch after 1.12.x branch has been made.

* Drop testing of Python 2.6, 3.2, and 3.3
* Create 1.12.0-notes.rst and add to source/documentation.
* Update pavement.py to use 1.10.x as LOG_START
* Update version numpy in setup.py

charris committed:
Merge pull request #7063 from charris/prepare-1.12-dev

REL: Update master branch after 1.12.x branch has been made.

jaimefrio committed:
Merge pull request #6997 from gfyoung/mtrand_helpers_compress

MAINT: Simplify mtrand.pyx helpers

charris committed:
DOC: Update the 1.11.0 release notes.

[ci skip]

charris committed:
MAINT: Update the git .mailmap file.

[ci skip]

Commit:
TST: Added broadcasting tests in test_random.py

Added a whole new suite of tests to ensure that
functions in mtrand.pyx which are broadcastable
actually broadcast their arguments properly.

Commit:
DOC: Clarified output size for broadcastable mtrand.pyx functions

Clarified the output size depending on whether
scalar or non-scalar inputs are passed to
functions in mtrand.pyx that can broadcast
their arguments.

charris committed:
Merge pull request #7076 from charris/update-mailmap

MAINT: Update the git .mailmap file.

xdmiodz committed:
BLD: fix compilation on non glibc-Linuxes

Non-glibc Linuxes dont have the __GLIBC_PREREQ function and compilation of numpy
fails on such platforms. To avoid this the TRIG_OK check should be done only in
the glibc environment

The patch is taken from AlpineLinux repository
http://git.alpinelinux.org/cgit/aports/tree/testing/py-numpy/numpy-1.10.0-musl.patch?id=2e5c4bfcf1c9746edd58a8e684d01403f234e71d

jakirkham committed:
BUG: In `norm`, always cast non-floating point arrays to 64-bit floats. Otherwise, weird integer roundoff errors give faulty results in some cases.

charris committed:
Merge pull request #7087 from xdmiodz/musl-build

BLD: fix compilation on non glibc-Linuxes

charris committed:
Revert "BLD: fix compilation on non glibc-Linuxes"

seberg committed:
DOC: Add documentation for as_strided

xdmiodz committed:
BLD: fix compilation on non glibc-Linuxes

Non-glibc Linuxes dont have the __GLIBC_PREREQ function and compilation of numpy
fails on such platforms.

charris committed:
Merge pull request #7091 from numpy/revert-7087-musl-build

Revert "BLD: fix compilation on non glibc-Linuxes"

charris committed:
Merge pull request #7092 from xdmiodz/glibc-check

BLD: fix compilation on non glibc-Linuxes

jakirkham committed:
TST: Verify that `norm` is properly casting values to floats as needed.

jakirkham committed:
BUG: Make sure that the `ord=0` case returns a float.

jakirkham committed:
DOC: Update `norm` docstring to include examples that reflect that all computations are done with floating point numbers.

charris committed:
Merge pull request #7088 from jakirkham/cast_float_linalg

BUG: Have `norm` cast non-floating point arrays to 64-bit float arrays

kawochen committed:
MAINT: Removed conditionals that are always true in datetime_strings.c

charris committed:
Merge pull request #7073 from charris/update-1.11.0-notes

DOC: Update the 1.11.0 release notes.

jaimefrio committed:
Merge pull request #7082 from gfyoung/test_random_expand

TST, DOC: Added Broadcasting Tests in test_random.py

alexbrc committed:
DOC: update Python versions requirements in the install docs

charris committed:
Merge pull request #7109 from argriffing/python-ver-dep-docs

DOC: update Python versions requirements in the install docs

jaimefrio committed:
Merge pull request #7055 from gfyoung/single_elt_array_return_matching

BUG: One Element Array Inputs Return Scalars in np.random

dongjoon-hyun committed:
MAINT: Fix typos in docs

jaimefrio committed:
Merge pull request #7114 from dongjoon-hyun/fix_typos_in_doc

MAINT: Fix typos in docs

Commit:
TST: Fixed f2py test for win32 virtualenv

Fixed test_scripts.test_f2py test so that
it can pass correctly on win32 virtualenvs,
in which the Python executable and the
f2py.py file are both in the Scripts
directory.

Commit:
TST: Fixed f2py test for non-versioned python executables

The 'sys.executable' can come in various names, but
the three main ones are "python", "python{major_version}",
and "python{major_version.minor_version}". The current
version of the f2py test assumes that only the latter
two are used. Since "f2py" is generally versioned,
using the executable basename "python" will make it
impossible to find. This commit fixes that issue
by using a sure-fire way of getting the Python version.

charris committed:
Merge pull request #7116 from gfyoung/f2py_patch

TST: Fixed f2py test for win32 virtualenv

Commit:
BUG: Fixed mingw.lib error

Closes gh-647.

charris committed:
Merge pull request #7118 from gfyoung/f2py_patch_linux

TST: Fixed f2py test for non-versioned python executables

madphysicist committed:
DOC: Updated documentation wording and examples for np.percentile.

Examples had some factual errors. Wording updated in a couple of places.

madphysicist committed:
MAINT: Accepted all review comments for PR#7125

madphysicist committed:
BUG: Fixed 'midpoint' interpolation of np.percentile in odd cases.

'midpoint' must return the same as 'higher' and 'lower' when the two
are the same, not 'lower' + 0.5 as it was doing.

Commit:
TST: test installing from sdist on TravisCI.

Commit:
BLD/BUG: revert usage of setuptools when building an sdist.

This was seriously broken.  Setuptools does unwanted 'smart' things and
ignores MANIFEST.in.  Closes gh-7127.

Commit:
MAINT: warn in egg_info command if using setuptools.sdist.

Warn in numpy/distutils/command/egg_info.py if using setuptools.sdist.

See gh-7127 for details.

charris committed:
Merge pull request #7131 from charris/fix-setuptools-sdist

Fix setuptools sdist

dongjoon-hyun committed:
MAINT: Fix some typos in a code string and comments

yarikoptic committed:
ENH: for savez create temporary file alongside with the target file

Closes: gh-5336

yarikoptic committed:
ENH: catch and rethrow exception in _savez upon IOError with filename info

homu committed:
Auto merge of #7134 - dongjoon-hyun:fix_typos_in_code_comments, r=njsmith

MAINT: Fix some typos in a code string and comments

yarikoptic committed:
BF: Since savez could be used with BytesIO - check if file is a string

rkern committed:
Merge pull request #7026 from gfyoung/uniform_bounds_bug

DOC: Clarify behavior in np.random.uniform

gmarkall committed:
BUG: Unpickled void scalars should be contiguous

Void scalars are both C- and Fortran-contiguous, so pickling and
unpickling them should result in a new void scalar that also has
these flags set. Fixes #7140.

gmarkall committed:
TST: Unpickled void scalars must be behaved

Unpickled void scalars must be both aligned and writeable.

juliantaylor committed:
Merge pull request #7141 from gmarkall/contig_void_unpickle

BUG: Unpickled void scalars should be contiguous

charris committed:
Merge pull request #7102 from kawochen/CLN-ISO-8601

MAINT: Removed conditionals that are always false in datetime_strings.c

shoyer committed:
Merge pull request #7125 from madphysicist/percentile-doc

DOC: Updated documentation wording and examples for np.percentile.

dongjoon-hyun committed:
MAINT: Change `call_fortran` into `callfortran` in comments.

It was committed mistakenly in #7134. `callfortran` is used a lot really.

njsmith committed:
Merge pull request #7144 from dongjoon-hyun/recover_callfortran

MAINT: Change `call_fortran` into `callfortran` in comments.

bastula committed:
BUG: Fixed regressions in np.piecewise in ref to #5737 and #5729.
Added unit tests for these conditions.

njsmith committed:
TST: remove duplicate test

There are two identical copies of test_dot_override in
test_multiarray.py. This seems surplus to requirements.

njsmith committed:
Temporarily disable __numpy_ufunc__

Given that we accidentally released 1.11b2 with `__numpy_ufunc__` still
enabled, we should probably just disable it in master for now.

When ready to re-enable, grep for NUMPY_UFUNC_DISABLED. Or just revert
this commit.

seberg committed:
TST: Add missing suffix to temppath manager

Without the suffix, np.save creates a new file and the file
does not get cleaned up.

mhvk committed:
Reascertain that linspace respects ndarray subclasses in start, stop.

charris committed:
Merge pull request #7149 from seberg/missing-suffix

TST: Add missing suffix to temppath manager

bertrand-l committed:
BUG: mode kwargs passed as unicode to np.pad raises an exception

isinstance(mode, str) is False in python2.7 when mode is of unicode
type, and mode is then mistakenly assumed to be a callable. See #7112

charris committed:
Merge pull request #7147 from njsmith/disable-numpy-ufunc

Temporarily disable __numpy_ufunc__

charris committed:
Merge pull request #7152 from bertrand-l/fix/unicode-kwarg-breaks-pad

BUG: mode kwargs passed as unicode to np.pad raises an exception

rgommers committed:
Merge pull request #7119 from gfyoung/mingw_patch

BUG: Fixed mingw.lib error

bastula committed:
MAINT: Addressed comments in PR #7145

charris committed:
Merge pull request #7156 from mhvk/linspace-subclass-safe2

Reascertain that linspace respects ndarray subclasses in start, stop.

charris committed:
Merge pull request #7145 from bastula/piecewise

BUG: Fixed regressions in np.piecewise in ref to #5737 and #5729.

charris committed:
Merge pull request #7129 from madphysicist/percentile-midpoint-interpolation

BUG: Fixed 'midpoint' interpolation of np.percentile in odd cases.

seberg committed:
Merge pull request #6656 from I--P/loadtxt-int

ENH: usecols now accepts an int when only one column has to be read

nbeaver committed:
DOC: base_repr *can* handle negative integers.

nbeaver committed:
TST: out of bounds bases in base_repr

charris committed:
Merge pull request #6298 from nbeaver/check-lower-base-limit

Check lower base limit in base_repr.

madphysicist committed:
Added 'doane' and 'sqrt' estimators to np.histogram in numpy.function_base

Commit:
Update Wikipedia references for mtrand.pyx

Specifically the 'Logarithmic distribution' link is fixed.
For others the links or article names are updated.

jaimefrio committed:
Merge pull request #7167 from 153957/patch-1

DOC: Update Wikipedia references for mtrand.pyx

madphysicist committed:
DOC: Updated minor typos in function_base.py and test_function_base.py

gfyoung committed:
TST: Fixed f2py test for Anaconda non-win32

When you run 'python -V' under Anaconda, it
returns for example, 'Python 3.4.3 :: Continuum
Analytics, Inc.' However, the original parsing
of the version in 'test_f2py' assumed there was
nothing following the version number, causing a
ValueError because you can't assign three
variables to four components that you get from
splitting on the '.'

charris committed:
Merge pull request #7171 from gfyoung/conda_f2py_patch

TST: Fixed f2py test for Anaconda non-win32

gfyoung committed:
DOC: Fix broken pandas link in 1.11.0 release notes

[ci skip]

charris committed:
Merge pull request #7174 from gfyoung/pandas_link_fix

DOC: Fix broken pandas link in release notes

madphysicist committed:
BUG: Fixed previous attempt to fix dimension mismatch in nanpercentile

nanpercentile was conforming to dimension convention of percentile incorrectly.
percentile outputs results for the different percentiles along the first
dimension of the output. nanpercentile was moving the reduction axis to the
front using swapaxes, which would move the first axis out of place if there
were more than two in the array. Added a test with more than two axes to
demonstrate and used rollaxis instead of swapaxes to do the interhange.

aukejw committed:
DOC: add vstack, hstack, dstack reference to stack documentation.

aukejw committed:
DOC: changed 'depth wise' to 'depthwise', 'the existing' to 'an existing'

madphysicist committed:
MAINT: Removed supurious assert in histogram estimators

jakirkham committed:
BUG: Raise a quieter `MaskedArrayFutureWarning` for mask changes.

Drop the `__getitem__` warning. In `__setitem__` check to see if the
masked array is shared. If it is shared, we know it will propagate
upstream in the future. Also, use a more specific warning type instead
of using `FutureWarning` so that this can be explicitly ignored.

jakirkham committed:
STY: Drop some trailing spaces in `numpy.ma.core`.

tacaswell committed:
STY: pep8 only

tacaswell committed:
MNT: move std, var, mean calls out of try block

Move the calls to user-provided versions of std, var, and mean on non
mu.ndarray objects out of the `try` block so that numpy will not mask
AttributeError raised during the execution of the function rather than
due to the object not having the required method.

tacaswell committed:
MNT: reduce number of return statements

In sum, amax, amin, and prod simplify the logic to remove an identical
return statement / call to `_methods._xxx`.  This removes several
elif/else pairs and reduces the number of exit points from the functions
but makes the code path a bit more complicated to trace.

charris committed:
Merge pull request #7195 from jakirkham/style_np_ma_fix

STY: Drop some trailing spaces in `numpy.ma.core`.

charris committed:
Merge pull request #7193 from madphysicist/extra-assert

MAINT: Removed supurious assert in histogram estimators

tacaswell committed:
DOC: add release notes

charris committed:
Merge pull request #7191 from aukejw/stack_documentation

DOC: add vstack, hstack, dstack reference to stack documentation.

shoyer committed:
Revert "DOC: add vstack, hstack, dstack reference to stack documentation."

jakirkham committed:
TST: Pin virtualenv used on Travis CI. [skip appveyor]

charris committed:
Merge pull request #7197 from jakirkham/pin_virtualenv

TST: Pin virtualenv used on Travis CI.

charris committed:
Merge pull request #7180 from madphysicist/nanpercentile-dims

BUG: Fixed previous attempt to fix dimension mismatch in nanpercentile

charris committed:
Merge pull request #7196 from numpy/revert-7191-stack_documentation

Revert "DOC: add vstack, hstack, dstack reference to stack documentation."

juliantaylor committed:
BUG: raise IOError on not a file in python2

The change in 5225e4c2007 did not account for PyFile_AsFile does not
raise an error on invalid input, add the handling to equalize our
wr5225e4c2007 apper function.
closes gh-7200

charris committed:
Merge pull request #7201 from juliantaylor/ioerror

Raise IOError on not a file in python2

charris committed:
Merge pull request #7181 from madphysicist/doc-typos

DOC: Updated minor typos in function_base.py and test_function_base.py

madphysicist committed:
MAINT: Made `iterable` return a boolean

charris committed:
Merge pull request #4619 from tacaswell/ndarray_subclass_sum_bug

BUG : np.sum silently drops keepdims for sub-classes of ndarray

jakirkham committed:
TST: Bump `virtualenv` to 14.0.6.

rgommers committed:
Merge pull request #7209 from jakirkham/use_virtualenv_14_0_6

TST: Bump `virtualenv` to 14.0.6

omegahm committed:
DOC: Fix fmin examples

This fixes #6904 (which has been closed already for some reason)

seberg committed:
Merge pull request #7211 from omegahm/doc-fix-fmin-examples

DOC: Fix fmin examples

jakirkham committed:
TST: Drop `FutureWarning` filters from tests where they were added.

njsmith committed:
Merge pull request #7133 from yarikoptic/bf-savez

savez: temporary file alongside with target file and improve exception message upon IOError

charris committed:
Merge pull request #7194 from jakirkham/fix_warning_masked_array_slices

BUG: Raise a quieter `MaskedArrayFutureWarning` for mask changes.

njsmith committed:
Merge pull request #7202 from madphysicist/iterable-bool-return

MAINT: Made `iterable` return a boolean

madphysicist committed:
DOC: Removed residual merge markup from previous commit

jaimefrio committed:
Merge pull request #7239 from madphysicist/patch-1

DOC: Removed residual merge markup from previous commit

building39 committed:
Change 'pubic' to 'public'.

rgommers committed:
Merge pull request #7240 from building39/feature/doc

DOC: typo: change 'pubic' to 'public'.

seberg committed:
Merge pull request #7090 from madphysicist/hist-estimators

ENH: Added 'doane' and 'sqrt' estimators to np.histogram in numpy.function_base

homu commented:
:umbrella: The latest upstream changes (presumably #7090) made this pull request unmergeable. Please resolve the merge conflicts.

seberg commented:
Wait, did we never put some kind of warning/error about using this with weights in, sounds like this should be done still!? @madphysicist maybe you want to have a bit of a look at this, or @nayyarv has time to finish this off?

rgommers committed:
DOC: update sphinxext to numpydoc 0.6.0

rgommers committed:
DOC: fix a number of reST formatting issues in docstrings.

rgommers committed:
DOC: fix up invalid LaTeX in histogram docstring.

madphysicist commented:
I skimmed this and the related (closed) PR. I will look into it in more detail as soon as I get a chance. The first thing that comes to mind is to add a `weightedfunctions` module similar to `nanfunctions`. However, even at my most optimistic, I have to admit this is *probably* not a reasonable solution.

charris committed:
Merge pull request #7241 from rgommers/update-sphinxext

MAINT: update doc/sphinxext to numpydoc 0.6.0, and fix up some docstrings

seberg commented:
Actually, I meant mostly if we are doing something reasonable with the
automatic bins right now. Error/warning ignore, I don't know, but just
to make sure it makes sense and is documented. We could try to make it
correct, but I expect it is not easy.

njsmith commented:
@seberg: Yeah, I think you're right -- it looks like in master and 1.11b3, the new `bins="auto"` feature is kinda broken: `np.histogram` will happily accept `bins="auto", weights=...`, and then return nonsense results. Similarly for `bins="auto", range=...`. This PR looks like a good solution to me, but has conflicts due to other cleanups to this code...

@nayyarv @madphysicist: Any chance one of you could clean this up quickly so we can get it into 1.11?

njsmith committed:
Merge pull request #7105 from seberg/writeable-as-strided

DEP: Deprecate as_strided returning a writable array as default

madphysicist commented:
I probably could some time this week, but certainly not over the next couple of days.

madphysicist commented:
Cancel my previous comment. I took a look at the code more carefully. I can clean this up right now actually. Can I just rebase it onto #7199?

nayyarv committed:
BUG: Change TypeError's raised by optimbins for weighted data to RunTimeWarnings instead.

nayyarv committed:
Merge remote-tracking branch 'origin/autobins' into autobins

# Conflicts:
#	numpy/lib/function_base.py
#	numpy/lib/tests/test_function_base.py

madphysicist commented:
I have rebased the `autobins` branch onto the latest `master` with all that that entails. I did add a couple of gramatical and line wrapping fixes. The result is PR #7243. I was unable to make a PR back to @nayyarv's fork of `numpy`.

@nayyarv, FWIW, I like the version in the `autobinsWeighted` branch better, even if this version is much simpler.

madphysicist commented:
Looks like the other PR kicked off a rebuild on this one?

nayyarv commented:
Nup, this was me trying to fix it, but looks like the rebase went a bit crazy.

madphysicist commented:
Looks like you may have rebased master onto your branch instead of the other way around?

madphysicist commented:
If you are OK with my PR, we can just run with that. I did not change any of the functionality, just some wording and whitespace stuff.

nayyarv commented:
yep, that's exactly what happened. Let's jump over to your PR instead, I'm shutting this down.

nayyarv commented:
Jump over to #7243, this pull request has gotten out of hand with my failure to git properly.

charris commented:
You can recover from a failed rebase, `git reflog` is your friend, See http://stackoverflow.com/questions/134882/undoing-a-git-rebase.

nayyarv commented:
Thanks charris, [I did manage to fix the rebase](https://github.com/numpy/numpy/compare/master...nayyarv:autobins) but since I closed this it, github won't let me reopen this Pull Request (since it's been recreated). As such, since madphysicist managed the exact same effect thing in his PR, I'm happy to let his go through instead
